File: 2025/python/solutions/day12.py
from lib.python import ProblemInput, ProblemResult, Solution

# Observation: All shapes are 3x3, almost fully filled.
# Thus, optimal packing is just filling as many full 3x3 blocks as possible,
# there is no room anyways for clever fitting.

SHAPE_SIZE = 3


class Solution12(Solution):
    @staticmethod
    def results() -> list[ProblemResult]:
        return [3, 448, None, None]

    @staticmethod
    def solve_part01(data: ProblemInput, is_sample: bool) -> ProblemResult:
        shape_defs, regions = parse(data)
        shape_sizes = [len(defn) for defn in shape_defs]

        valid, invalid, unsure = 0, 0, 0
        for (width, height), shape_count in regions:
            region_area = width * height
            raw_tile_count = sum(c * s for c, s in zip(shape_count, shape_sizes))
            min_region_required = (SHAPE_SIZE * SHAPE_SIZE) * sum(shape_count)

            if min_region_required <= region_area:
                # Simple square tiling is already enough
                valid += 1
            elif raw_tile_count > region_area:
                # More tiles than available
                invalid += 1
            else:
                # Could fit, but only with clever tiling
                unsure += 1

        # Assuming all unsure work out (is == 0 for input anyways)
        assert is_sample or unsure == 0
        return valid + unsure

    @staticmethod
    def solve_part02(data: ProblemInput, is_sample: bool) -> ProblemResult:
        return None


def parse(
    lines: ProblemInput,
) -> tuple[list[list[tuple[int, int]]], list[tuple[tuple[int, int], list[int]]]]:
    shape_defs = []
    shape_counts = []

    i = 0
    while i < len(lines):
        line = lines[i].strip()

        # Start of shape definition
        if len(line) == 2 and line.endswith(":"):
            i += 1
            shape = []
            for y in range(SHAPE_SIZE):
                for x in range(SHAPE_SIZE):
                    if lines[i + y][x] == "#":
                        shape.append((x, y))
            shape_defs.append(shape)
            i += SHAPE_SIZE + 1

        # Area definition
        elif ":" in line and "x" in line.split(":")[0]:
            dimensions, present_counts = line.split(":", 1)
            width, height = map(int, dimensions.strip().split("x"))
            counts = list(map(int, present_counts.strip().split()))

            shape_counts.append(((width, height), counts))
            i += 1

    return shape_defs, shape_counts

File: 2025/elixir/solutions/day12.ex
defmodule Day12 do
  @moduledoc """
  Solution for Advent of Code 2025 - Day 12
  """

  @behaviour AocLib.Solution

  alias AocLib.Types.ProblemInput

  @impl true
  def results do
    [
      :unsolved,
      448,
      :no_part_two,
      :no_part_two
    ]
  end

  @impl true
  def solve_part01(input, is_sample) do
    {shape_sizes, region_requirements} = input |> ProblemInput.lines() |> parse()

    {valid, _invalid, unsure} =
      for {{width, height}, present_counts} <- region_requirements, reduce: {0, 0, 0} do
        {valid, invalid, unsure} ->
          region_area = width * height

          raw_tile_count =
            Enum.zip(present_counts, shape_sizes) |> Enum.sum_by(fn {c, s} -> c * s end)

          min_region_required = 3 * 3 * Enum.sum(present_counts)

          cond do
            # Put each present in its own 3x3 area, so definitely valid
            min_region_required <= region_area ->
              {valid + 1, invalid, unsure}

            # Even if we cut all present in unit tiles, this wont fit, so definitely invalid
            raw_tile_count > region_area ->
              {valid, invalid + 1, unsure}

            # Can only fit with clever packing
            true ->
              {valid, invalid, unsure + 1}
          end
      end

    if not is_sample and unsure == 0, do: valid, else: :unsolved
  end

  @impl true
  def solve_part02(_input, _is_sample) do
    :no_part_two
  end

  defp parse(lines) do
    {regions, shapes} =
      lines
      |> Enum.chunk_by(&(&1 == ""))
      |> Enum.reject(&(length(&1) == 1 and hd(&1) == ""))
      |> List.pop_at(-1)

    shape_sizes =
      shapes
      |> Enum.map(fn shape ->
        tl(shape) |> Enum.join() |> String.graphemes() |> Enum.count(&(&1 == "#"))
      end)

    region_requirements =
      regions
      |> Enum.map(fn def ->
        [dimensions, present_counts] = String.split(def, ": ", parts: 2, trim: true)

        [width, length] = dimensions |> String.split("x") |> Enum.map(&String.to_integer/1)
        presents = present_counts |> String.split(" ") |> Enum.map(&String.to_integer/1)
        {{width, length}, presents}
      end)

    {shape_sizes, region_requirements}
  end
end
